### PR TITLE
Node location cache: off-heap storage and “array” implementation that supports parallel inserts

### PR DESCRIPTION
**Pull request by msbarry**
Add more options for node location storage (aka `LongLongMap` since lat/lon gets packed into a 64-bit long for storage) and some code improvements along the way:

- `--nodemap-type=array` option that stores node locations in an array indexed by node ID (8 bytes * max node ID) - and also supports parallel inserts. This makes OSM pass 1 2-3x faster on larger machines
- `--nodemap-storage=direct` option to store node location cache in native off-heap "direct" memory using `ByteBuffer.allocateDirect` (you'll likely need to increase the `-XX:MaxDirectMemorySize` setting to use this)
- Extracted common resource use checking code into `ResourceUsage` class that prints a more detailed breakdown of usage
- Switched to enums to represent long long map and storage types
- Extracted worker coordination logic (i.e. finish all nodes before moving onto ways) and element counting into an `OsmPhaser` class that also logs more details when a node begins/ends
- Switch to `posix_madvise` instead of `madvise`

land checklist:
- [x] Add off-heap storage option for existing implementations using `ByteBuffer.allocateDirect`
- [x] Add `"array"` nodemap type that supports parallel inserts from multiple threads
- [x] Refactor `osm_pass1` phase to support a sequential and parallel execution model, based on if the node map implementation supports parallel inserts
- [x] Include off-heap (direct/mmap) usage during resource usage checks at startup
- [x] Add enums for nodemap storage and implementations
- [x] Memory usage estimation
- [x] Array mmap cleanup (is there a better way to do sliding window limit?)
- [x] Are primitive arrays any faster than HeapByteBuffer getLong/putLong? If not then simplify to one ByteBuffer implementation that switches whether it uses `allocate` vs. `allocateDirect`
- [x] Check direct memory limit `-XX: MaxDirectMemorySize` if set, otherwise `Runtime.getRuntime().maxMemory()`.
- [x] When nodes/ways/relations done, print # per second (and a summary at the end of osm pass1/2)

### CODE REVIEW

**Comment by bchapuis** [comment]
Are you noticing a significant gains with the parallel version? I guess that the machine used to produce the benchmarks displayed in the issue has a limited number of cores.

When experimenting with your code to replace lmdb in baremaps, I tried to make the memory and the serialization of records pluggable. The resulting `AlignedDataList` class is almost identical to your `AppendStore` and illustrates this relatively well.  The constructor takes two parameters: an `AlignedDataType` (a fixed size data type aligned to the memory); a 'Memory' (which can be on-heap, off-heap, and memory-mapped). 

https://github.com/baremaps/baremaps/blob/8cabb72922480fbe7d6b2ad8130a2704b52bf210/baremaps-collection/src/main/java/com/baremaps/collection/AlignedDataList.java#L51

As the `Memory` interface solely rely on `ByteBuffer`s it is relatively easy to provide the on-heap, off-heap and on-disk implementations with `byteBuffer.allocate`, `byteBuffer.allocateDirect`, and `channel.map`. When it comes to `DataType`s, the idea is to solely rely on absolute positions to access the content of the buffer to enable zero-copy serialization and deserialization. I guess that you don't need this for your use case, but I would be really intereseted to know what you think about it.

**Comment by msbarry** [comment]
> Are you noticing a significant gains with the parallel version? I guess that the machine used to produce the benchmarks displayed in the issue has a limited number of cores.

Yes, the parallel version is quite a bit faster (for writes at least) - but only when running over the whole planet. I ran on a machine with 32GB RAM and 16 cpus, and the memory-mapped sparse array took about 11 minutes (15-17 million nodes per second, 4-6 million ways per second) but the parallel array implementation took 4m44s (40-50 million nodes per second, 6-10 million ways per second).  And the data storage ends up only a little bit larger (74 instead of 68gb). The RAM parallel array implementation only takes 4m10s.

`AlignedDataList` looks pretty close to this! Once writes are positional though, they no longer have to be sequential can happen from parallel threads. I'm curious how the write performance is on memory-mapped byte buffers? In general I've seen that reads from memory mapped byte buffers are fast, but writes are slower than other mechanisms (some more details [here](https://mapdb.org/blog/mmap_files_alloc_and_jvm_crash/)).  

I am also curious if there's any noticeable difference using `getLong`/`putLong` on a `HeapByteBuffer` instead of `int[]` or `long[]`? My assumption was that regular arrays would be fastest, but if there's no difference then it would be nice to just have one `ByteBuffer` implementation that changes by whether it calls `allocate` or `allocateDirect`.

**Comment by bchapuis** [comment]
I just tried to run a [benchmark](https://github.com/baremaps/baremaps/blob/feature/benchmarks/baremaps-benchmarks/src/main/java/com/baremaps/jmh/WriteMemoryBenchmark.java) to answer the question about write performance. Most of the benchmark I saw online were quite outdated (4 years and older).

Writing data on-disk with mapped buffers looks suspiciously fast on my 2020 macbook pro (I use the force method to ensure that the changes are written to the storage device).

When it comes to byte arrays and heap buffers, an overhead can be noticed when writing less than 1mb. Above 1mb, the two solutions look in par.

Direct buffers are slower than heap buffers, but the garbage collector is probably less solicited.

```
Benchmark                                          (bytes)  Mode  Cnt      Score      Error  Units
WriteMemoryBenchmark.writeBufferedOutputStream        1000    ss    3      0.508 ±    0.051  ms/op
WriteMemoryBenchmark.writeBufferedOutputStream       10000    ss    3      0.747 ±    3.422  ms/op
WriteMemoryBenchmark.writeBufferedOutputStream      100000    ss    3      3.093 ±    5.460  ms/op
WriteMemoryBenchmark.writeBufferedOutputStream     1000000    ss    3     23.331 ±   15.014  ms/op
WriteMemoryBenchmark.writeBufferedOutputStream    10000000    ss    3    221.102 ±   20.761  ms/op
WriteMemoryBenchmark.writeBufferedOutputStream   100000000    ss    3   2162.074 ±  669.249  ms/op
WriteMemoryBenchmark.writeBufferedOutputStream  1000000000    ss    3  21576.192 ±  857.302  ms/op
WriteMemoryBenchmark.writeByteArray                   1000    ss    3      0.018 ±    0.042  ms/op
WriteMemoryBenchmark.writeByteArray                  10000    ss    3      0.156 ±    0.343  ms/op
WriteMemoryBenchmark.writeByteArray                 100000    ss    3      0.369 ±    1.183  ms/op
WriteMemoryBenchmark.writeByteArray                1000000    ss    3      2.074 ±   26.183  ms/op
WriteMemoryBenchmark.writeByteArray               10000000    ss    3      8.898 ±   38.439  ms/op
WriteMemoryBenchmark.writeByteArray              100000000    ss    3     26.375 ±   46.770  ms/op
WriteMemoryBenchmark.writeByteArray             1000000000    ss    3    262.262 ± 1033.857  ms/op
WriteMemoryBenchmark.writeByteBuffer                  1000    ss    3      0.050 ±    0.031  ms/op
WriteMemoryBenchmark.writeByteBuffer                 10000    ss    3      0.345 ±    0.617  ms/op
WriteMemoryBenchmark.writeByteBuffer                100000    ss    3      0.639 ±    2.152  ms/op
WriteMemoryBenchmark.writeByteBuffer               1000000    ss    3      2.402 ±   44.364  ms/op
WriteMemoryBenchmark.writeByteBuffer              10000000    ss    3      9.275 ±   46.315  ms/op
WriteMemoryBenchmark.writeByteBuffer             100000000    ss    3     28.598 ±   78.618  ms/op
WriteMemoryBenchmark.writeByteBuffer            1000000000    ss    3    274.356 ±  883.777  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer            1000    ss    3      0.064 ±    0.214  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer           10000    ss    3      0.469 ±    2.310  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer          100000    ss    3      0.745 ±    1.025  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer         1000000    ss    3      2.355 ±   42.834  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer        10000000    ss    3      8.460 ±   62.737  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer       100000000    ss    3     61.445 ±   85.125  ms/op
WriteMemoryBenchmark.writeDirectByteBuffer      1000000000    ss    3    793.005 ± 3241.072  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer          1000    ss    3      0.436 ±    0.708  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer         10000    ss    3      0.653 ±    1.584  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer        100000    ss    3      1.446 ±    0.669  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer       1000000    ss    3      3.465 ±   56.955  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer      10000000    ss    3     18.684 ±  135.192  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer     100000000    ss    3    146.181 ±   71.446  ms/op
WriteMemoryBenchmark.writeMemoryMappedBuffer    1000000000    ss    3   1332.760 ± 1238.829  ms/op
```

Do not hesitate to let me know if you spot an issue in the benchmark. For now, the result suggest that using `ByteBuffer`s would be a good solution as they could be used to access on-heap, off-heap and on-disk memorier.

**Comment by msbarry** [comment]
Thanks @bchapuis! I did some testing too with `long[]` vs. `ByteBuffer.putLong()` and they seem to be close enough, so I just dropped the `int[]` and `long[]` storage options for the append store and array implementations.

**Comment by msbarry** [comment]
Ran some performance tests on a c5ad.16xlarge (64 cpu, 128gb ram) instance:

tl;dr) The parallel array nodemap implementation shaves about 5 minutes off of OSM pass 1 (8->3m) when storage=ram or 4 minutes when storage=mmap (8->4m).  Also node read performance in pass 2 is about the same when RAM or MMAP storage is used, so switching to MMAP and using a smaller Xmx setting (~45g or less) will only add about 1 minute to generation time.

**Xmx=115g nodemap=array/ram**
- full planet generation took 47m8s (cpu:26h53m3s gc:3m46s avg:34.2) - full logs in planet-logs/v0.3.0-planet-c5ad-128gb.txt
- osm_pass1 took 2m58s (cpu:1h57m gc:35s avg:39.4)
- osm_pass2 took 21m49s (cpu:19h14m46s gc:2m18s avg:52.9)

```
0:03:53 DEB [osm_pass1] - Processed 18,767 blocks:
0:03:53 INF [osm_pass1] -   nodes: 7,546,573,682 (108M/s) in 1m9s cpu:58m23s gc:10s avg:50.5
0:03:53 INF [osm_pass1] -   ways: 841,994,745 (8.5M/s) in 1m39s cpu:55m56s gc:24s avg:34
0:03:53 INF [osm_pass1] -   relations: 9,718,612 (1.1M/s) in 9s cpu:1m20s avg:9.3
0:03:53 INF [osm_pass1] - Finished in 2m58s cpu:1h57m gc:35s avg:39.4
0:03:53 INF [osm_pass1] -   read     1x(32% 58s sys:35s wait:1m11s done:9s)
0:03:53 INF [osm_pass1] -   process 63x(38% 1m8s sys:3s block:1m10s wait:8s)

0:25:42 DEB [osm_pass2] - Processed 18,767 blocks:
0:25:42 INF [osm_pass2] -   nodes: 7,546,573,682 (26M/s) in 4m45s cpu:4h4m2s gc:37s avg:51.4
0:25:42 INF [osm_pass2] -   ways: 841,994,745 (971k/s) in 14m27s cpu:12h32m24s gc:1m33s avg:52.1
0:25:42 INF [osm_pass2] -   relations: 9,718,612 (61k/s) in 2m38s cpu:2h39m41s gc:8s avg:60.6
0:25:42 INF [osm_pass2] - Finished in 21m49s cpu:19h14m46s gc:2m18s avg:52.9
0:25:42 INF [osm_pass2] -   read     1x(4% 50s sys:31s wait:17m48s done:2m40s)
0:25:42 INF [osm_pass2] -   process 63x(72% 15m40s block:51s wait:2m15s)
0:25:42 INF [osm_pass2] -   write    1x(64% 13m52s sys:1m16s wait:5m40s)
```

**Xmx=115g nodemap=sparsearray/ram**
- osm_pass1 took 8m10s (cpu:2h22m54s gc:1m6s avg:17.5)
- (killed during pass2, but it was progressing at about the same rate)

```
0:09:07 DEB [osm_pass1] - Processed 18,767 blocks:
0:09:07 INF [osm_pass1] -   nodes: 7,546,573,682 (23M/s) in 5m18s cpu:1h10m24s gc:39s avg:13.3
0:09:07 INF [osm_pass1] -   ways: 841,994,745 (5.2M/s) in 2m42s cpu:1h11m16s gc:26s avg:26.4
0:09:07 INF [osm_pass1] -   relations: 9,718,612 (1M/s) in 9s cpu:33s avg:3.7
0:09:07 INF [osm_pass1] - Finished in 8m10s cpu:2h22m54s gc:1m6s avg:17.5
0:09:07 INF [osm_pass1] -   read     1x(6% 31s sys:21s wait:7m8s done:7s)
0:09:07 INF [osm_pass1] -   parse   62x(8% 40s wait:7m16s done:7s)
0:09:07 INF [osm_pass1] -   process  1x(86% 7m1s)
```

**Xmx=45g nodemap=array/mmap/madvise=true**
- osm_pass1 took 4m11s (cpu:1h48m24s gc:30s avg:26)
- psm_pass2 took 21m41s (cpu:19h59m35s gc:2m8s avg:55.3)

```
0:05:08 DEB [osm_pass1] - Processed 18,767 blocks:
0:05:08 INF [osm_pass1] -   nodes: 7,546,573,682 (53M/s) in 2m20s cpu:45m4s gc:9s avg:19.3
0:05:08 INF [osm_pass1] -   ways: 841,994,745 (8.2M/s) in 1m41s cpu:1h1m30s gc:21s avg:36.4
0:05:08 INF [osm_pass1] -   relations: 9,718,612 (1.1M/s) in 8s cpu:1m12s avg:8.7
0:05:08 INF [osm_pass1] - Finished in 4m11s cpu:1h48m24s gc:30s avg:26
0:05:08 INF [osm_pass1] -   read     1x(17% 42s sys:25s wait:1m57s done:10s)
0:05:08 INF [osm_pass1] -   process 63x(23% 57s sys:2s block:1m17s wait:1m32s)

0:26:50 DEB [osm_pass2] - Processed 18,767 blocks:
0:26:50 INF [osm_pass2] -   nodes: 7,546,573,682 (28M/s) in 4m27s cpu:3h51m59s gc:30s avg:52.2
0:26:50 INF [osm_pass2] -   ways: 841,994,745 (962k/s) in 14m35s cpu:13h29m22s gc:1m30s avg:55.5
0:26:50 INF [osm_pass2] -   relations: 9,718,612 (60k/s) in 2m42s cpu:2h39m26s gc:8s avg:59.2
0:26:50 INF [osm_pass2] - Finished in 21m41s cpu:19h59m35s gc:2m8s avg:55.3
0:26:50 INF [osm_pass2] -   read     1x(4% 49s sys:31s wait:17m33s done:2m43s)
0:26:50 INF [osm_pass2] -   process 63x(76% 16m23s sys:8s block:43s wait:10s)
0:26:50 INF [osm_pass2] -   write    1x(55% 11m58s sys:1m13s wait:7m39s)
```

**Comment by bchapuis** [comment]
MMap's reading performences are impressive. Is your use of madvise significantly impact performence or do you think something changed in the way memory mapped files are implemented in Java 17?

**Comment by msbarry** [comment]
> MMap's reading performances are impressive. Is your use of madvise significantly impact performance or do you think something changed in the way memory mapped files are implemented in Java 17?

Before Java 13 you needed to set the position on byte buffers then read in separate calls, then Java 13 added positional get methods so multiple threads can read from the same file without synchronization so I think that helped use cases like this. Now as long as you have more free memory than the file, it should be almost as fast as reading from RAM. Madvise probably didn't make much of a difference here, but it should help when the amount of free ram is less than the node cache file.

**Comment by bchapuis** [comment]
Thanks a lot for the clarification, I was aware about the positional get methods, but I would not have thought that it would have such an impact, even if a lot of ram is available.
